### PR TITLE
Fix COPY data types issues

### PR DESCRIPTION
**Pull request by Jesus89**
Fixes https://github.com/CartoDB/cartoframes/issues/784.

This PR contains important fixes related to uploading and downloading data types :tada:

- **Fix upload types**: the old implementation it was iterating by the row. To get a row, panda unifies the type of all the columns in the row, so, if for example, you have a column with `int` and another with `float` it assumes that the row it type `float` so it will add your `int` column data as `float` causing a non-deterministic bug. Now the iteration is made using the index of the row, so the final type of the value is the correct one.
- **Encode numeric types NaN, Infinity, -Infinity**: tt adds a map between `numpy` types np.nan, np.inf and -np.inf.
- **Fix `dtypes` to `pgtypes` and `pgtypes` to `dtypes` conversion**: using PostgreSQL types instead: https://www.postgresql.org/docs/7.4/datatype.html. Using `pgtype` allows a better mapping for the DataFrame types. The PG types aliases have been added too (int4, float8, etc).
- **Allow upload NULL values**: in the old implementation, null values were ignored or converted to ' ' . In this PR NaN and None valued in the DataFrame are properly converted to `NaN` and `null` in the DB.
- **Fix download NULL and Infinite values**: it allows to download `null`, `NaN`, `Infinite`, `-Infinite` values from the CARTO table.
- **Fix upload coordinates (0, 0)**: there is also a small fix to enable uploading the geometry `POINT(0 0)`.

### CODE REVIEW

File: cartoframes/data/dataset/registry/dataframe_dataset.py [hunk 1]
```diff
@@ -164,9 +164,9 @@ def _rows(df, dataframe_columns_info, with_lnglat):
             row_data.append(encode_row(val))
 
         if with_lnglat:
-            lng_val = row[with_lnglat[0]]
-            lat_val = row[with_lnglat[1]]
-            if lng_val and lat_val:
+            lng_val = df[with_lnglat[0]][index]
```
Comment (andy-esch) [inline cartoframes/data/dataset/registry/dataframe_dataset.py]: `df.at` may be a more appropriate access method for this type of use case: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.at.html

File: cartoframes/data/dataset/registry/dataframe_dataset.py [hunk 2]
```diff
@@ -139,20 +141,18 @@ def _is_valid_index_for_cartodb_id(index):
...
-                    val = i
+                    val = index
                 else:  # we could have filtered columns in the df. See DataframeColumnsInfo
                     continue
             else:
-                val = row[col]
-
-            if _is_null(val):
-                val = ''
+                val = df[col][index]
```
Comment (andy-esch) [inline cartoframes/data/dataset/registry/dataframe_dataset.py]: Same as below with [`df.at`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.at.html)

File: cartoframes/data/dataset/registry/dataframe_dataset.py (hunk 1 above)
Comment (Jesus89) [reply cartoframes/data/dataset/registry/dataframe_dataset.py]: Cool thanks

File: cartoframes/data/dataset/registry/dataframe_dataset.py (hunk 2 above)
Comment (Jesus89) [reply cartoframes/data/dataset/registry/dataframe_dataset.py]: :+1:

**Comment by andy-esch** [comment]
❤️Love this one! Bugs like these are maybe the most annoying! Beyond the small refactor suggestion, I don't see any recommended changes. Most of it is over my head unless I spend more time thinking through it.

File: cartoframes/utils/utils.py [hunk 3]
```diff
@@ -325,6 +347,17 @@ def remove_column_from_dataframe(dataframe, name):
 
 
 def encode_row(row):
+    if row is None:
+        row = PG_NULL
+
+    if isinstance(row, float):
```
Comment (alrocar) [inline cartoframes/utils/utils.py]: Should we use `elseif`?

File: cartoframes/utils/columns.py [hunk 4]
```diff
@@ -87,6 +94,12 @@ def _slugify(self, value):
         return value
 
 
+def _extract_pgtype(fields):
+    if 'pgtype' in fields:
+        return fields['pgtype']
+    return None
```
[the remark this answers is not in the document]
Comment (Jesus89) [reply cartoframes/utils/columns.py]: In this case, pgtype is None, so the default dtype is assigned ('object')

**Comment by Jesus89** [comment]
We have just unit tests for that. I'll add a couple of e2e tests.

File: cartoframes/utils/utils.py (hunk 3 above)
Comment (Jesus89) [reply cartoframes/utils/utils.py]: Yep

**Comment by alrocar** [comment]
Acceptance 🍏

**Pull request merged**